src/ORM/EagerLoadable.php

```diff
@@ -0,0 +1,268 @@
+    protected $_forMatching;
+
+    /**
+     * Constructor. The $config parameted accepts the following array
+     * keys:
+     *
+     * - associations
+     * - instance
+     * - config
+     * - canBeJoined
+     * - aliasPath
+     * - propertyPath
+     * - forMatching
+     *
+     * The keys maps to the settable properties in this class.
+     *
+     * @param string $name The Association name.
+     * @param array $config The list of properties to set.
+     */
+    public function __construct($name, array $config = [])
+    {
+        $this->_name = $name;
+        $allowed = [
+            'associations', 'instance', 'config', 'canBeJoined',
+            'aliasPath', 'propertyPath', 'forMatching'
+        ];
+        foreach ($allowed as $property) {
+            if (isset($config[$property])) {
+                $this->{'_' . $property} = $config[$property];
+            }
+        }
+    }
+
+    /**
+     * Adds a new association to be loaded from this level.
+     *
+     * @param string $name The association name.
+     * @param \Cake\ORM\EagerLoadable $association The association to load.
+     * @return void
+     */
+    public function addAssociation($name, EagerLoadable $association)
+    {
+        $this->_associations[$name] = $association;
+    }
+
+    /**
+     * Returns the Association class instance to use for loading the records.
+     *
+     * @return array
+     */
+    public function associations()
+    {
+        return $this->_associations;
+    }
+
+    /**
+     * Sets the Association class instance to use for loading the records.
+     *
+     * If called with no arguments it returns the current
+     * value.
+     *
+     * @param \Cake\ORM\Association|null $instance The value to set.
+     * @return \Cake\ORM\Association|null
+     */
+    public function instance($instance = null)
+    {
+        if ($instance === null) {
+            return $this->_instance;
+        }
+        $this->_instance = $instance;
+    }
+
+    /**
+     * Sets a dotted separated string representing the path of associations
+     * that should be followed to fetch this level.
+     *
+     * If called with no arguments it returns the current
+     * value.
+     *
+     * @param string|null $path The value to set.
+     * @return string|null
+     */
+    public function aliasPath($path = null)
+    {
+        if ($path === null) {
+            return $this->_aliasPath;
+        }
+        $this->_aliasPath = $path;
+    }
+
+    /**
+     * Sets a dotted separated string representing the path of entity properties
+     * in which results for this level should be placed.
+     *
+     * If called with no arguments it returns the current
+     * value.
+     *
+     * @param string|null $path The value to set.
+     * @return string|null
+     */
+    public function propertyPath($path = null)
+    {
+        if ($path === null) {
+            return $this->_propertyPath;
+        }
+        $this->_propertyPath = $path;
```

Comment:
Is there any reason for these properties to be mutable? Outside of canBeJoined & config none of the properties are ever mutated. It might be simpler to have a value object.